src/v/cluster/types.cc

```diff
@@ -298,11 +305,22 @@ void adl<cluster::topic_configuration>::to(
       t.properties.timestamp_type,
       t.properties.segment_size,
       t.properties.retention_bytes,
-      t.properties.retention_duration);
+      t.properties.retention_duration,
+      t.properties.recovery,
+      t.properties.shadow_indexing);
 }
 
 cluster::topic_configuration
 adl<cluster::topic_configuration>::from(iobuf_parser& in) {
+    auto version = adl<int32_t>{}.from(in.peek(4));
+    if (version < 0) {
+        // Consume version from stream
+        vassert(
```

Comment:
Cool solution! So your pretty much checking that expected field (size of ns) is negative or not. The only way it could be negative would be if it was serialized in the new format. Nice.